### DISCUSSION TITLE
Allow Users to view/edit all entries of their User Group

### DISCUSSION THREAD

**Discussion by denisyilmaz** [craftcms/cms#12745] [Ideas]
Would it be possible to extend the user group rule set by an option that defines if a user group can see/edit entries of their own user group?

We have a bigger project with 18 different departments with multiple users who should only get access to their pages that they belong to and furthermore see the hierachy of their entries. Currently, the only way to solve this requirement with built-in features would be to create a section per department.

As the pages of the departments don't differ from their blueprint this means we need to duplicate the field layout for all 18 departments. Additionally, as we intent to use Craft CMS in a headless way we furthermore need to repeat the field fragments for all section types for a simple general `entry(uri:"example-uri") {title ... on section_default_Entry {that_field}}` query.
Somehow this feels way to complicated for a simple "allow user X to access entries which belong to user group X" requirement.

With the new conditions logic that would allow to show/hide fields per `userGroup` and the recent update to have a "viewable/saveble" conditions #12266  that will be introduces in 4.4 it seems that adding this user role setting would fit in that approach.

Are we missing any built-in feature that might have the solution for our requirement?

**Comment by brandonkelly** [comment]
Interesting idea! For now, you could pull this off by granting “other author” permissions, and then scaling that back with a custom event handler in a module:

```php
use craft\elements\Entry;
use craft\events\AuthorizationCheckEvent;
use craft\services\Elements;
use yii\base\Event;

Event::on(Elements::class, Elements::EVENT_AUTHORIZE_VIEW, function(AuthorizationCheckEvent $event) {
    if (
        $event->element instanceof Entry &&
        $event->element->section->handle === 'mySectionHandle' &&
        $event->element->author &&
        $event->element->author->id !== $event->user->id
    ) {
        // See if the author shares at least one user group with the current user
        $hasCommonGroup = false;
        foreach ($event->element->author->getGroups() as $group) {
            if ($event->user->isInGroup($group)) {
                $hasCommonGroup = true;
                break;
            }
        }
        if (!$hasCommonGroup) {
            $event->authorized = false;
        }
    }
});
```

All entries will then be listed in element indexes, but they won’t get a hyperlink to the edit page, if the current user doesn’t share any groups with the entry’s author.

**Reply by denisyilmaz** [reply]
Thanks for the reply. This solution is only disabling the hyperlink to the edit page, right? It wouldn't for example prevent users from moving entries within the structure of the overview page? This brings me to another possible addition to the user roles: enable/disable structural changes to entries in a structure.

In general we would find it very useful to have more granular control over which user is capable of which CP features. Like:
- can add/edit entry types X and Y but not Z within section A
- can move entries in structure (would also prevent "can change parent of entry")
- can change status of entries (enable/disable for sites)

These are a bit outside of the scope of this discussion here, but would be interesting to know if there are any plans to extend the user-roles logic in craftcms 4?
